extensions-core/lookups-cached-global/src/main/java/io/druid/server/lookup/namespace/cache/NamespaceExtractionCacheManager.java

```diff
@@ -412,23 +431,79 @@ public void run()
   }
 
   /**
-   * This method is expected to swap the cacheKey into the active namespace, and leave future requests for new cacheKey available. getCacheMap(cacheKey) should return empty data after this call.
+   * This method swap the cacheKey into the active namespace, and leave future requests for new cacheKey available. getCacheMap(cacheKey) should return empty data after this call.
    *
    * @param namespaceKey The namespace to swap the cache into
    * @param cacheKey     The cacheKey that contains the data of interest
    *
    * @return true if old data was cleared. False if no old data was found
    */
-  protected abstract boolean swapAndClearCache(String namespaceKey, String cacheKey);
+  protected boolean swapAndClearCache(String namespaceKey, String cacheKey)
+  {
+    final Lock lock = nsLocks.get(namespaceKey);
+    lock.lock();
+    try {
+      ConcurrentMap<String, Map<String, String>> cacheMap = mapMap.get(cacheKey);
+      if (cacheMap == null) {
+        throw new IAE("Extraction Cache [%s] does not exist", cacheKey);
+      }
+      ConcurrentMap<String, Map<String, String>> prior = mapMap.put(namespaceKey, cacheMap);
+      mapMap.remove(cacheKey);
+      if (prior != null) {
+        // Old map will get GC'd when it is not used anymore
+        return true;
+      } else {
+        return false;
+      }
+    }
+    finally {
+      lock.unlock();
+    }
+  }
 
   /**
    * Return a ConcurrentMap with the specified ID (either namespace's name or a cache key ID)
+   * Create empty map if no matching map exists
    *
    * @param namespaceOrCacheKey Either a namespace or cache key should be acceptable here.
    *
-   * @return A ConcurrentMap<String, String> that is backed by the impl which implements this method.
+   * @return A ConcurrentMap<String, Map<String, String>>
+   */
+  public ConcurrentMap<String, Map<String, String>> getCacheMap(String namespaceOrCacheKey)
+  {
+    ConcurrentMap<String, Map<String, String>> map = mapMap.get(namespaceOrCacheKey);
+
+    if (map == null) {
+      mapMap.putIfAbsent(namespaceOrCacheKey, new ConcurrentHashMap<String, Map<String, String>>());
+      map = mapMap.get(namespaceOrCacheKey);
+    }
+    return map;
+  }
+
+  /**
+   * Return a ConcurrentMap with the specified ID and mapName
+   * Create empty map if no matching mapName exists
+   *
+   * @param id A namespace key should be acceptable here.
+   *
+   * @param mapName A name of map in a given namespace
+   *
+   * @return A Map<String, String>
    */
-  public abstract ConcurrentMap<String, String> getCacheMap(String namespaceOrCacheKey);
+  public Map<String, String> getCacheMap(String id, String mapName)
```

Comment:
let's rename this to something like getCacheInnerMap() to differentiate the two functions, and note in the javadocs that this retrieves an inner map